collector/utils.go

```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		return err
+	}
+
+	c.cgroupfs, err = dc.GetDockerParam("CgroupDriver")
+	if err != nil {
+		return err
+	}
+
+	drv, err := dc.GetDockerParam("Driver")
+	if err != nil {
+		return err
+	}
+
+	c.client = dc
+	c.driver = drv
+
+	log.WithFields(log.Fields{
+		"plugin": "docker_collector",
+		"block":  "initClient",
+	}).Infof("Docker client initialized with %s and %s drivers", drv, c.cgroupfs)
+
+	return nil
+}
+
+// createMetricNamespace returns metric namespace based on given `ns` which is used as a prefix; all dynamic elements
+// in the `metricName` are defined based on content of map `dynamicElements`
+func (creator *nsCreator) createMetricNamespace(ns plugin.Namespace, metricName string) (plugin.Namespace, error) {
+	metricName = strings.TrimSpace(metricName)
+
+	if len(metricName) == 0 {
+		return nil, fmt.Errorf("Cannot create metric namespace: empty metric name %s", metricName)
+	}
+
+	elements := strings.Split(metricName, "/")
+
+	// check if metricName contains only static elements
+	if !strings.Contains(metricName, "*") {
+		ns = ns.AddStaticElements(elements...)
+		return ns, nil
+	}
+
+	// when metric name contains dynamic element iterate over elements
+	for index, element := range elements {
+		if element == "*" {
+			// the following element is dynamic
+			dynamicElement, ok := creator.dynamicElements[elements[index-1]]
+			// check if this dynamic element is supported (name and description are available)
+			if !ok {
+				return nil, fmt.Errorf("Unknown dynamic element in metric `%s` under index %d", metricName, index)
+			}
+			// add recognize dynamic element (define its name and description)
+			ns = ns.AddDynamicElement(dynamicElement.name, dynamicElement.description)
+
+			if len(elements)-1 == index {
+				// in case when an asterisk is the last element, add `value` at the end of ns
+				ns = ns.AddStaticElement("value")
+			}
+		} else {
+			// the following element is static
+			ns = ns.AddStaticElement(element)
+		}
+	}
+	if len(ns) == 0 {
+		return nil, fmt.Errorf("Cannot create metric namespace for metric %s", metricName)
+	}
+	return ns, nil
+}
+
+func getQueryGroup(ns []string) (string, error) {
+	if ns[0] == "spec" {
+		return ns[0], nil
+	}
+
+	for _, ne := range ns {
+		if _, exists := getters[ne]; exists {
+			return ne, nil
+		}
+	}
+	return "", fmt.Errorf("Cannot identify query group for given namespace %s", strings.Join(ns, "/"))
+}
+
+func appendIfMissing(collectGroup map[string]map[string]struct{}, rid string, query string) {
+	group, exists := collectGroup[rid]
+	if !exists {
+		collectGroup[rid] = map[string]struct{}{query: struct{}{}}
+	}
+
+	if _, exists := group[query]; !exists {
+		collectGroup[rid][query] = struct{}{}
+	}
+}
+
+func getDockerConfig(cfg plugin.Config) (map[string]string, error) {
+	config := make(map[string]string)
+	values := []string{"endpoint", "procfs"}
+	var err error
+	for _, v := range values {
+		config[v], err = getStringFromConfig(cfg, v)
```

Comment:
Could you explain why `cfg.GetString(key)` is not called directly, but a new function `getStringFromConfig` was used for do it?